### ISSUE TITLE
Downloaded Database Dump JSON data format conversion failures

### ISSUE THREAD

**Issue by neilt** [internetarchive/openlibrary#2369]
### Description
First, I am not sure this is the correct place to report this.  If not, please provide some guidance.

1. I downloaded the open library data full OL dump from the web page [https://openlibrary.org/developers/dumps](https://openlibrary.org/developers/dumps). Note this is NOT the complete dump.

2. Then I imported the data into a PostgreSQL database version 11.05. The JSON field was imported as text.

3. Then I ran a PostgreSQL function `is_valid_json(cte."json")`  defined below on each row. JSON failed validation for 7001152 out of 53586852 rows, or about 13%. Most failures occurred in the work, edition, and authors data. If desired I can provide exact info down the the key field for each row..

4. Figuring this might be a JSON conversion issue with the PostgreSQL code I then wrote a quick Apple Swift program to examine each row in the database and convert the text to JSON. This code was essentially:
```swift
    _ = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
```
This Apple Swift JSON conversion failed for the exact same rows as the PostgreSQL function did. So I now believe that the failures are in the data supplied by OpenLibrary.  For the Apple Swift conversion failures I can provide in addition to the row key the exact type of failure for each row and the character position.

The initial failures were unescaped 0a and 0d control characters. These appeared to be in text I did not care about so I simply removed them. The next round the failures are malformed JSON, with an example error of `Badly formed object around character 1478`.  At this point I have not investigated further.

At this point it appears to me, based on my limited investigation, that the code producing the data dumps is allowing invalid JSON to be created.  This could be occurring because the code producing the dumps or the data in the database.

My questions are:

A. Is this a known issue with the data. I searched the open GitHub issues and did not find one, but I could have easily missed it.

B. Is there a process for reporting and correcting these data issues.  If so, please point me to it.

I am quit happy to spend some time helping define the problems if someone or some process is going to correct the original data.

if it would be useful to someone I can also provide much more detailed information like the row key, the character position of the error, and the type of error.  However it does seem like these errors are the type that should be removed by scanning the original data source or by code, not by hand.

Please let me know how to proceed.

### Evidence / Screenshot (if possible)
None

### Relevant url?
[https://openlibrary.org/developers/dumps](https://openlibrary.org/developers/dumps)

### Expectation
I expect the JSON to be valid JSON or if it is some funky version of JSON to have that documented.

### Details

PG Database Version: 11.05
PG DB Build Version: PostgreSQL 11.5 on x86_64-apple-darwin18.7.0, compiled by Apple LLVM version 10.0.1 (clang-1001.0.46.4), 64-bit

ProductName:	Mac OS X
ProductVersion:	10.14.6
BuildVersion:	18G95

PostgreSQL function to validate JSON:
```
        CREATE OR REPLACE FUNCTION is_valid_json(p_json text) RETURNS boolean AS $$
        BEGIN
          return (p_json::jsonb is not null);
        exception
          when others then
             return false;
        END;
        $$
        language plpgsql immutable;
```

PostgreSQL query to check and track JSON validation for a block of rows.
```
        WITH cte AS (
                SELECT "key", "json"
                FROM   open_lib_import
                WHERE  pg_json_valid IS NULL OR pg_json_valid = FALSE
                LIMIT  10000
        )
        UPDATE open_lib_import AS i
        SET    pg_json_valid = is_valid_json(cte."json")
        FROM   cte
        WHERE  i."key" = cte."key"
```

**Comment by neilt** [comment]
Here is an example:
```
Type: /type/edition, Key: /books/OL12315664M, Rev: 6, Last: 2012-06-24 17:41:37.364243-05
Error: Unescaped control character around character 261.
  243 2c , 
  244 20   
  245 22 " 
  246 76 v 
  247 61 a 
  248 6c l 
  249 75 u 
  250 65 e 
  251 22 " 
  252 3a : 
  253 20   
  254 22 " 
  255 52 R 
  256 65 e 
  257 76 v 
  258 69 i 
  259 73 s 
  260 65 e 
  261 64 d  <--
  262 0d 
 
  263 0a 
 
  264 33 3 
  265 72 r 
  266 64 d 
  267 20   
  268 45 E 
  269 64 d 
  270 69 i 
  271 74 t 
  272 69 i 
  273 6f o 
  274 6e n 
  275 22 " 
  276 7d } 
  277 2c , 
  278 20   
  279 22 " 
  280 69 i 
  281 64 d 
```

**Comment by hornc** [comment]
Hi @neilt , thanks for looking at our Open Library data and for reporting the problem you found. Especially for the examples!

Answering your specific questions from above:

A. This is not a known issue. Our JSON in those fields _should_ be valid standard JSON for all records. If it's not, we'll need to make changes!

B. This is the correct place to report issues with data and code. Github makes it convenient to tack all issues in one place, and data issues tend to eventually have a corresponding code fix .

 Thanks for the example, /books/OL12315664M 's JSON can be viewed directly here: https://openlibrary.org/books/OL12315664M.json   My browser (Chrome + JSON viewer) hides the new line and carriage return character, turning it into a space, but when I view source I can see the `"value": "Revised\r\n3rd Edition"`  `\r\n`  characters in the `"notes"` field. So they are there in the data and in the individual JSON representation, as well as the dump files.

I checked, and `\r\n` are valid escaped characters in JSON https://tools.ietf.org/html/rfc4627#section-2.5

and the JSON itself passes validation: https://jsonformatter.org/3179e2

What I think might be happening is that because you imported the JSON field as text, the escaped `\r\n` were not processed as JSON escaped characters. I'm not sure the exact interface you are using, but Postgres has had native JSON support since 9.2, so there should be a way to import that data into a json typed column.

I think that covers the specific example above, but I'm worried your report of "unescaped 0a and 0d control characters" is a problem in the data, and I wouldn't be too surprised if there are a range of character problems, so I'm not sure we are out of the woods yet, and appreciate your report!

I'll look into other potential invalid JSON issues in the data, and am happy to work with you to get the data imported correctly into your db. Let me know whether it is possible to import that as `json` rather than `text`, and if that at least reduces the number of errors you see.

**Comment by tfmorris** [comment]
You don't say what dump you tried, but I've successfully loaded both the 2019-08-31 and 2019-007-31 dumps into PostgreSQL as JSONB. The only thing which needed to be filtered (on a very tiny fraction of records) is embedded NULs (ie \u0000).

**Comment by cdrini** [comment]
I've been able to successfully load the full dump for 2019-08-31 (i.e. https://openlibrary.org/data/ol_dump_latest.txt.gz ) into postgres as JSONB as well. Here's the code I used:

https://github.com/cdrini/openlibrary/blob/1401cc47c1e080ffd6047a9ef1e0a902b38af13a/scripts/solr_builder/psql-import-chunk.sh#L7-L11

and here's the table schema:

https://github.com/cdrini/openlibrary/blob/1401cc47c1e080ffd6047a9ef1e0a902b38af13a/scripts/solr_builder/sql/create-dump-table.sql#L1-L7

Which dump file were you trying to access? There might be an error in one of the other ones.

**Comment by neilt** [comment]
I originally tried to load directly into a PostgreSQL jsonb field and it failed. That is the reason I tried loading the json data into a text field.  I loaded the data using PgAdmin 4 import, but I am sure PgAdmin 4 did not use the options that were shown by cdrini above. I should have known better as I don't trust anything else PgAdmin 4 does.

I used the same table schema shown by cdrini above when it initially failed using jsonb field.

The download dump I used was at [https://openlibrary.org/data/ol_dump_latest.txt.gz](https://openlibrary.org/data/ol_dump_latest.txt.gz). Which appears to be the same as others have used. Not sure which day I downloaded, but I'll try again with the latest.

With this information, let me try again.

Thanks for the excellent help.  I'll post back, either working or not.

**Comment by cdrini** [comment]
> I should have known better as I don't trust anything else PgAdmin 4 does.

😄 

Let use know if it works :)

**Comment by tfmorris** [comment]
> The download dump I used was at https://openlibrary.org/data/ol_dump_latest.txt.gz. ... Not sure which day I downloaded, but I'll try again with the latest.

If you use `wget --trust-server-names` or equivalent, you'll get the actual filename that was downloaded, which is always useful information.

**Comment by neilt** [comment]
Well I downloaded [ol_dump_2019-08-31.txt.gz](ol_dump_2019-08-31.txt.gz) and it imported into a jsonb column just fine using `psql`.  Sorry for the noise.

Apparently the problem was PgAdmin 4 mangling the data.

For others, here is the macOS command used to import the data into Postgres. `sed` was a little different and there were more columns in the table, otherwise just as cdrini suggested.
```
zcat ol_dump_2019-08-31.txt.gz | \
sed 's/\\u0000//g' | \
psql -d new_books -c "COPY open_lib_import ("key", "type", revision, last_modified, json_bin) FROM STDIN with delimiter E'\t' escape '\' quote E'\b' csv"
```

Thanks again for the great support.

**Comment by tfmorris** [comment]
Glad you got it sorted out.

**Comment by neilt** [comment]
No reason to keep this open now, so closing.

**Comment by tfmorris** [comment]
@mekarpeles The above suspicious link should be deleted. I've reported it to Github, but it shouldn't be left here to tempt people to click on it.

**Comment by cdrini** [comment]
Thank you @tfmorris I deleted the comment.
